### ISSUE TITLE
inlinable function call in a function with debug info must have a !dbg location

### ISSUE THREAD

**Issue by nathanchance** [ClangBuiltLinux/linux#1510]
The kernel test robot [reported](https://lore.kernel.org/r/202111200929.CBA2BzkH-lkp@intel.com/):

```
$ curl https://lore.kernel.org/all/202111200929.CBA2BzkH-lkp@intel.com/2-a.bin | gzip -d > .config

$ make -skj"$(nproc)" LLVM=1 LLVM_IAS=1 olddefconfig vmlinux
inlinable function call in a function with debug info must have a !dbg location
  call void @__tsan_write8(i8* %3)
inlinable function call in a function with debug info must have a !dbg location
  call void @__tsan_write8(i8* %5)
inlinable function call in a function with debug info must have a !dbg location
  call void @__tsan_read8(i8* %4)
inlinable function call in a function with debug info must have a !dbg location
  call void @__tsan_read8(i8* %7)
inlinable function call in a function with debug info must have a !dbg location
  call void @__tsan_read8(i8* %16)
...
```

Initially, I thought that this required two files to reproduce, I chose `kernel/time/timekeeping.c` and `kernel/kcsan/core.c`. With that, `cvise` spits out:

```
$ cat timekeeping.i
int tk_normalize_xtime_tk_0_0;
long tk_normalize_xtime_tk_0_1;
void timekeeping_suspend() {
  while (tk_normalize_xtime_tk_0_1 >= tk_normalize_xtime_tk_0_0)
    ;
}

$ cat core.i
void __tsan_read8(void *ptr) {}

$ clang -fshort-wchar -O2 -flto -g -fno-stack-protector -c -o core.{o,i}

$ clang -fshort-wchar -O2 -flto -g -fsanitize=thread -c -o timekeeping.{o,i}

$ llvm-ar cDPrST built-in.a core.o timekeeping.o

$ ld.lld -m elf_x86_64 -r -o /dev/null --whole-archive built-in.a
inlinable function call in a function with debug info must have a !dbg location
  call void @__tsan_read8(i8* bitcast (i64* @tk_normalize_xtime_tk_0_1 to i8*))
LLVM ERROR: Broken module found, compilation aborted!
...
```

but I can reproduce it without LTO when `__tsan_read8` is in the translation unit:

```
$ cat {core,timekeeping}.i > test.i

$ cat test.i
void __tsan_read8(void *ptr) {}
int tk_normalize_xtime_tk_0_0;
long tk_normalize_xtime_tk_0_1;
void timekeeping_suspend() {
  while (tk_normalize_xtime_tk_0_1 >= tk_normalize_xtime_tk_0_0)
    ;
}

$ clang -fshort-wchar -O2 -g -fsanitize=thread -c -o /dev/null test.i
inlinable function call in a function with debug info must have a !dbg location
  call void @__tsan_read8(i8* bitcast (i64* @tk_normalize_xtime_tk_0_1 to i8*))
inlinable function call in a function with debug info must have a !dbg location
  call void @__tsan_read8(i8* bitcast (i64* @tk_normalize_xtime_tk_0_1 to i8*))
fatal error: error in backend: Broken module found, compilation aborted!
...
```

Initially, I did not think this was a regression but it was reproducible with LLVM 12 and not LLVM 11. Bisecting turned up https://github.com/llvm/llvm-project/commit/dfc5a9eb57aaaac972764bf731503419284bd3dc:

```
$ git bisect run ...
...
dfc5a9eb57aaaac972764bf731503419284bd3dc is the first bad commit
commit dfc5a9eb57aaaac972764bf731503419284bd3dc
Author: Vedant Kumar <vsk@apple.com>
Date:   Tue Sep 8 13:39:52 2020 -0700

    [Instruction] Add dropLocation and updateLocationAfterHoist helpers

    Introduce a helper which can be used to update the debug location of an
    Instruction after the instruction is hoisted. This can be used to safely
    drop a source location as recommended by the docs.

    For more context, see the discussion in https://reviews.llvm.org/D60913.

    Differential Revision: https://reviews.llvm.org/D85670

 llvm/include/llvm/IR/Instruction.h                 | 14 +++++
 llvm/lib/IR/DebugInfo.cpp                          | 32 ++++++++++
 llvm/lib/Transforms/Scalar/GVN.cpp                 |  4 +-
 llvm/lib/Transforms/Scalar/LICM.cpp                |  5 +-
 .../test/DebugInfo/Generic/licm-hoist-debug-loc.ll |  3 +-
 llvm/test/Transforms/GVN/PRE/phi-translate.ll      |  7 +--
 .../Transforms/LICM/hoisting-preheader-debugloc.ll |  2 +-
 llvm/unittests/IR/InstructionsTest.cpp             | 72 ++++++++++++++++++++++
 8 files changed, 125 insertions(+), 14 deletions(-)
bisect found first bad commit

$ git bisect log
# bad: [5369517d20dd362a178a1b2d6c398d8898ee4620] Bump the trunk major version to 13
# good: [7ab7b979d29e1e43701cf690f5cf1903740f50e3] Bump the trunk major version to 12
git bisect start 'llvmorg-13-init' 'llvmorg-12-init'
# bad: [7579941d8b02af995e152a453da9b9cc0cc205e7] [examples] Fix Kaleidoscope-Ch3, which was broken by ad92f16ccc5.
git bisect bad 7579941d8b02af995e152a453da9b9cc0cc205e7
# good: [deb4b2580715810ecd5cb7eefa5ffbe65e5eedc8] [DAGCombine] Don't delete the node if it has uses immediately
git bisect good deb4b2580715810ecd5cb7eefa5ffbe65e5eedc8
# good: [270d334a665faa574db0c7d3a23af78bed9366d0] [docs][llvm] Fix typos
git bisect good 270d334a665faa574db0c7d3a23af78bed9366d0
# skip: [27e1cc6f391b57c9e20344c8a6d77a57f697eb87] Cleanup CodeGen/CallingConvLower.cpp
git bisect skip 27e1cc6f391b57c9e20344c8a6d77a57f697eb87
# skip: [1127662c6dc2a276839c75a42238b11a3ad00f32] [SelectionDAG] Make sure FMF are propagated when getSetcc canonicalizes FP constants to RHS.
git bisect skip 1127662c6dc2a276839c75a42238b11a3ad00f32
# bad: [8f56e382f782632c4b137339ab3c362fdbab62a5] [SCEV] Do not apply info from loop guards in AddRecs.
git bisect bad 8f56e382f782632c4b137339ab3c362fdbab62a5
# bad: [4ab45cc2260d87f18e1b05517d5d366b2e754b72] [AArch64][GlobalISel] Add some more legal types for G_PHI, G_IMPLICIT_DEF, G_FREEZE.
git bisect bad 4ab45cc2260d87f18e1b05517d5d366b2e754b72
# bad: [645c53a9d923c8d7f6a4b49aa02126aec23dc667] [ValueTracking] enhance isKnownNeverInfinity to understand sitofp
git bisect bad 645c53a9d923c8d7f6a4b49aa02126aec23dc667
# skip: [9ed1e5873c19eb817fb9e36d0262c7effee5d35e] [mlir][shape] Start a pass that lowers shape constraints.
git bisect skip 9ed1e5873c19eb817fb9e36d0262c7effee5d35e
# good: [99edb9b79c46706a41921c4edb1b849c1a618d26] [flang][driver] Fix options flag in the frontend driver
git bisect good 99edb9b79c46706a41921c4edb1b849c1a618d26
# bad: [6f10998ad80687a85633d8dda84146c57c49ca25] [flang][driver] Add missing dependency (shared library builds, NFC)
git bisect bad 6f10998ad80687a85633d8dda84146c57c49ca25
# good: [e34bd1e0b03d20a506ada156d87e1b3a96d82fa2] [APFloat] prevent NaN morphing into Inf on conversion (PR43907)
git bisect good e34bd1e0b03d20a506ada156d87e1b3a96d82fa2
# good: [506b6170cb513f1cb6e93a3b690c758f9ded18ac] Reland [CodeGen] emit CG profile for COFF object file
git bisect good 506b6170cb513f1cb6e93a3b690c758f9ded18ac
# bad: [a32feed0dbeac7606d042d0d7e041c9eaf12cd51] Add a static_assert confirming that DiagnosticBuilder is small
git bisect bad a32feed0dbeac7606d042d0d7e041c9eaf12cd51
# bad: [070555c6c008c57c408c1e99025b12c5d71c9848] [lld] Make -z keep-text-section-prefix recognize .text.split. as a prefix.
git bisect bad 070555c6c008c57c408c1e99025b12c5d71c9848
# good: [1c5a3c4d382353e582ecf4913e338599028e267f] [WebAssembly] Make SjLj lowering globals thread-local
git bisect good 1c5a3c4d382353e582ecf4913e338599028e267f
# bad: [c32e69b2ce7abfb151a87ba363ac9e25abf7d417] [lld-macho][re-land] Initial support for common symbols
git bisect bad c32e69b2ce7abfb151a87ba363ac9e25abf7d417
# bad: [dfc5a9eb57aaaac972764bf731503419284bd3dc] [Instruction] Add dropLocation and updateLocationAfterHoist helpers
git bisect bad dfc5a9eb57aaaac972764bf731503419284bd3dc
# first bad commit: [dfc5a9eb57aaaac972764bf731503419284bd3dc] [Instruction] Add dropLocation and updateLocationAfterHoist helpers
```

I intend to report this upstream but it seems like [Bugzilla is being migrated to GitHub this week](https://lists.llvm.org/pipermail/llvm-dev/2021-November/153802.html) so I will wait until after that is done.

**Comment by emojifreak** [comment]
The same (?) error is reproduced with DEBUG_INFO=y + LTO + allmodconfig as below. The below log is for compiling arm64 5.15.5 kernel. Compiling environment is Debian unstable amd64 with LLVM13:

```
  LTO     vmlinux.o
inlinable function call in a function with debug info must have a !dbg location
  call void @__sanitizer_cov_trace_const_cmp4(i32 32763, i32 %41)
inlinable function call in a function with debug info must have a !dbg location
  call void @__sanitizer_cov_trace_const_cmp4(i32 0, i32 %23)
inlinable function call in a function with debug info must have a !dbg location
  call void @__sanitizer_cov_trace_const_cmp4(i32 -1, i32 %41)
inlinable function call in a function with debug info must have a !dbg location
  call void @__sanitizer_cov_trace_const_cmp4(i32 0, i32 %53)
inlinable function call in a function with debug info must have a !dbg location
  call void @__sanitizer_cov_trace_const_cmp4(i32 0, i32 %81)
inlinable function call in a function with debug info must have a !dbg location
  call void @__sanitizer_cov_trace_const_cmp4(i32 -1, i32 %18)
LLVM ERROR: Broken module found, compilation aborted!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.	Program arguments: ld.lld -EL -maarch64elf -z norelro -mllvm -import-instr-limit=5 -r -o vmlinux.o -T .tmp_initcalls.lds -T .tmp_symversions.lds --whole-archive arch/arm64/kernel/head.o init/built-in.a usr/built-in.a arch/arm64/built-in.a kernel/built-in.a certs/built-in.a mm/built-in.a fs/built-in.a ipc/built-in.a security/built-in.a crypto/built-in.a block/built-in.a arch/arm64/lib/built-in.a lib/built-in.a arch/arm64/lib/lib.a lib/lib.a drivers/built-in.a sound/built-in.a samples/built-in.a net/built-in.a virt/built-in.a --no-whole-archive --start-group ./drivers/firmware/efi/libstub/lib.a --end-group
 #0 0x00007f781aee7431 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xde9431)
 #1 0x00007f781aee55e0 llvm::sys::RunSignalHandlers() (/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xde75e0)
 #2 0x00007f781aee7932 (/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xde9932)
 #3 0x00007f78200bd8e0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x138e0)
 #4 0x00007f7819c0ae71 raise (/lib/x86_64-linux-gnu/libc.so.6+0x3ce71)
 #5 0x00007f7819bf4536 abort (/lib/x86_64-linux-gnu/libc.so.6+0x26536)
 #6 0x00007f781ae32e99 (/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xd34e99)
 #7 0x00007f781ae32cb6 (/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xd34cb6)
 #8 0x00007f781b0797fc (/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xf7b7fc)
 #9 0x00007f781c1d5c2d (/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x20d7c2d)
#10 0x00007f781b04f2fe llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module> >::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/lib/x86_64-linux-gnu/libLLVM-13.so.1+0xf512fe)
#11 0x00007f781c1d0d9c llvm::lto::opt(llvm::lto::Config const&, llvm::TargetMachine*, unsigned int, llvm::Module&, bool, llvm::ModuleSummaryIndex*, llvm::ModuleSummaryIndex const*, std::vector<unsigned char, std::allocator<unsigned char> > const&) (/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x20d2d9c)
#12 0x00007f781c1d12c7 llvm::lto::backend(llvm::lto::Config const&, std::function<std::unique_ptr<llvm::lto::NativeObjectStream, std::default_delete<llvm::lto::NativeObjectStream> > (unsigned int)>, unsigned int, llvm::Module&, llvm::ModuleSummaryIndex&) (/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x20d32c7)
#13 0x00007f781c1c6c04 llvm::lto::LTO::runRegularLTO(std::function<std::unique_ptr<llvm::lto::NativeObjectStream, std::default_delete<llvm::lto::NativeObjectStream> > (unsigned int)>) (/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x20c8c04)
#14 0x00007f781c1c641b llvm::lto::LTO::run(std::function<std::unique_ptr<llvm::lto::NativeObjectStream, std::default_delete<llvm::lto::NativeObjectStream> > (unsigned int)>, std::function<std::function<std::unique_ptr<llvm::lto::NativeObjectStream, std::default_delete<llvm::lto::NativeObjectStream> > (unsigned int)> (unsigned int, llvm::StringRef)>) (/lib/x86_64-linux-gnu/libLLVM-13.so.1+0x20c841b)
#15 0x0000000000620913 lld::elf::BitcodeCompiler::compile() (/usr/lib/llvm-13/bin/lld+0x620913)
#16 0x00000000005ac7e6 void lld::elf::LinkerDriver::compileBitcodeFiles<llvm::object::ELFType<(llvm::support::endianness)1, true> >() (/usr/lib/llvm-13/bin/lld+0x5ac7e6)
#17 0x0000000000598c5c void lld::elf::LinkerDriver::link<llvm::object::ELFType<(llvm::support::endianness)1, true> >(llvm::opt::InputArgList&) (/usr/lib/llvm-13/bin/lld+0x598c5c)
#18 0x000000000058e1ed lld::elf::LinkerDriver::linkerMain(llvm::ArrayRef<char const*>) (/usr/lib/llvm-13/bin/lld+0x58e1ed)
#19 0x000000000058c420 lld::elf::link(llvm::ArrayRef<char const*>, bool, llvm::raw_ostream&, llvm::raw_ostream&) (/usr/lib/llvm-13/bin/lld+0x58c420)
#20 0x00000000004d77e5 (/usr/lib/llvm-13/bin/lld+0x4d77e5)
#21 0x00000000004d70e5 main (/usr/lib/llvm-13/bin/lld+0x4d70e5)
#22 0x00007f7819bf5e4a __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x27e4a)
#23 0x00000000004d6daa _start (/usr/lib/llvm-13/bin/lld+0x4d6daa)
Aborted
make[1]: *** [Makefile:1183: vmlinux] Error 134
make: *** [Makefile:1162: autoksyms_recursive] Error 2
  LDS     scripts/module.lds
  GEN     scripts/gdb/linux/constants.py
make: Target 'all' not remade because of errors.
```

Reproduction script is below:

```
#!/bin/sh

KVER=5.15.5
LANG=C.UTF-8
export LANG
cd /var/tmp/tmp
wget -T 10 -c https://cdn.kernel.org/pub/linux/kernel/v5.x/linux-${KVER}.tar.xz

  cat >/tmp/config$$.txt <<EOF
CONFIG_FTRACE_MCOUNT_USE_RECORDMCOUNT=n
CONFIG_KASAN=n
CONFIG_GCOV_KERNEL=n
CONFIG_COMPILE_TEST=n
CONFIG_LTO_CLANG_FULL=y
CONFIG_TRIM_UNUSED_KSYMS=y
CONFIG_LD_DEAD_CODE_DATA_ELIMINATION=y
CONFIG_CFI_CLANG=y
CONFIG_CFI_CLANG_SHADOW=y
CONFIG_SHADOW_CALL_STACK=y
CONFIG_INIT_STACK_ALL_ZERO=y
CONFIG_PREEMPT=y
CONFIG_PREEMPT_DYNAMIC=y

CONFIG_WERROR=n
CONFIG_XFS_FS=n
CONFIG_OVERLAY_FS=n
CONFIG_EMBEDDED=n

CONFIG_DEBUG_INFO=y
CONFIG_DEBUG_INFO_COMPRESSED=n
CONFIG_DEBUG_INFO_DWARF5=y
CONFIG_GDB_SCRIPTS=y

EOF


for a in arm64 x86 i386; do
  rm -rf /var/tmp/tmp/$a
  mkdir /var/tmp/tmp/$a
  cd /var/tmp/tmp/$a
  exec </dev/null >build-log-${KVER}-${a}.txt 2>&1
  set -xe
  tar Jxf /var/tmp/tmp/linux-${KVER}.tar.xz
  cd linux-${KVER}

  cp /tmp/config$$.txt .config
  make LLVM=1 LLVM_IAS=1 allmodconfig
  cat /tmp/config$$.txt >>.config

  
  yes '' |
    make  --keep-going ARCH=$a LLVM=1 LLVM_IAS=1 all &
  exec </dev/null >/dev/null 2>&1
done 
wait
```

**Comment by nathanchance** [comment]
Yes, I suspect that is going to be the same root cause because it is the same message with a sanitizer (in this case, KCOV).

**Comment by nathanchance** [comment]
This was already reported upstream, I just added on this information there: https://github.com/llvm/llvm-project/issues/47220

**Comment by nathanchance** [comment]
Huh, I was combing through old reports and I noticed that @nickdesaulniers fixed an instance of this for KCOV in https://github.com/ClangBuiltLinux/linux/issues/1338, which I think should have fixed the instance that @emojifreak reported?

**Comment by emojifreak** [comment]
I can still consistently reproduce this error with linux 5.16-rc5 and

```
$ clang --version
Debian clang version 13.0.0-9+b2
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by emojifreak** [comment]
Thin LTO + allyesconfig + DEBUG_INFO=y also produces this.

**Comment by nathanchance** [comment]
For KCSAN: https://reviews.llvm.org/D124937

**Comment by nickdesaulniers** [comment]
I think all of the issues are fixed here; there are two conflating reports.

The initial report is specifically KCSAN. https://reviews.llvm.org/D124937 just landed and will ship in clang-15.

The instance from @emojifreak I fixed back in https://reviews.llvm.org/D100158, which shipped as 4914c9836765532ffaec9c85bc57690dbdaeb50f in release/13.x.
